File: backend/app/schemas/extraction.py
from datetime import datetime

from pydantic import Field, field_validator, model_validator

from app.schemas.base import ApiModel as BaseModel


class ExtractionJobOut(BaseModel):
    id: str
    session_id: str
    transcript_version_id: str
    status: str
    provider: str
    model: str
    extractor_version: str
    prompt_version: str
    generation_no: int = 1
    supersedes_job_id: str | None = None
    retry_count: int
    max_retries: int
    error_code: str | None
    error_message: str | None
    created_at: datetime
    started_at: datetime | None
    completed_at: datetime | None
    review_lock_user_id: str | None = None
    review_lock_acquired_at: datetime | None = None
    review_lock_expires_at: datetime | None = None
    model_config = {"from_attributes": True}


class ExtractionBatchRerunIn(BaseModel):
    session_ids: list[str] = Field(min_length=1, max_length=50)

    @field_validator("session_ids")
    @classmethod
    def unique_session_ids(cls, values: list[str]) -> list[str]:
        normalized = [value.strip() for value in values if value.strip()]
        if not normalized:
            raise ValueError("至少选择一条测评任务记录")
        if len(set(normalized)) != len(normalized):
            raise ValueError("测评任务记录不能重复选择")
        return normalized


class ExtractionBatchRerunItemOut(BaseModel):
    session_id: str
    status: str
    message: str
    job: ExtractionJobOut | None = None


class ExtractionBatchRerunOut(BaseModel):
    requested: int
    created: int
    skipped: int
    failed: int
    items: list[ExtractionBatchRerunItemOut]


class ExtractionJobStatusOut(BaseModel):
    id: str
    session_id: str
    status: str
    generation_no: int
    retry_count: int
    max_retries: int
    candidate_count: int = 0
    error_code: str | None = None
    error_message: str | None = None
    created_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None


class ExtractionJobStatusBatchIn(BaseModel):
    job_ids: list[str] = Field(min_length=1, max_length=50)

    @field_validator("job_ids")
    @classmethod
    def unique_job_ids(cls, values: list[str]) -> list[str]:
        normalized = [value.strip() for value in values if value.strip()]
        if not normalized:
            raise ValueError("至少需要查询一个抽取任务")
        if len(set(normalized)) != len(normalized):
            raise ValueError("抽取任务不能重复")
        return normalized


class ExtractionJobStatusBatchOut(BaseModel):
    items: list[ExtractionJobStatusOut]


class ExtractionCandidateOut(BaseModel):
    id: str
    extraction_job_id: str
    source_transcript_segment_id: str | None
    sequence_no: int
    source_type: str
    review_status: str
    raw_asr_text: str
    original_text: str
    clean_text: str
    char_start: int | None
    char_end: int | None
    started_at_ms: int
    ended_at_ms: int
    reviewer_id: str | None
    review_note: str
    reviewed_at: datetime | None
    updated_at: datetime
    is_low_risk: bool = False
    classifier_version: str | None = None
    predicted_label: int | None = None
    predicted_dimension: str | None = None
    prediction_confidence: float | None = None
    prediction_probabilities: dict | None = None
    classified_at: datetime | None = None
    classification_error: str = ""
    classification_status: str = "pending_classification"
    prediction_source: str | None = None
    model_config = {"from_attributes": True}


class ExtractionQueueItemOut(BaseModel):
    session_id: str
    run_id: str | None
    user_id: str
    user_name: str
    username: str
    class_group: str | None
    task_id: str
    task_title: str
    sequence_no: int
    completed_at: datetime | None
    completed_at_source: str
    transcript_version_no: int | None
    transcript_source: str | None
    asr_status: str
    asr_error_code: str | None
    asr_error_message: str | None
    audio_available: bool
    job: ExtractionJobOut | None
    candidate_count: int
    pending_count: int


class ExtractionQueuePageOut(BaseModel):
    items: list[ExtractionQueueItemOut]
    total: int
    page: int
    page_size: int
    total_pages: int
    class_groups: list[str]
    tasks: list[dict]
    statuses: list[str]


class TranscriptEvidenceSegmentOut(BaseModel):
    id: str
    segment_no: int | None
    text: str
    started_at_ms: int
    ended_at_ms: int
    model_config = {"from_attributes": True}


class ExtractionReviewDetailOut(BaseModel):
    session_id: str
    run_id: str | None
    user_id: str
    user_name: str
    username: str
    task_id: str
    task_title: str
    sequence_no: int
    transcript_version_id: str | None
    transcript_version_no: int | None
    transcript_source: str | None
    full_text: str
    audio_available: bool
    asr_status: str
    asr_error_code: str | None
    asr_error_message: str | None
    job: ExtractionJobOut | None
    job_history: list[ExtractionJobOut] = Field(default_factory=list)
    segments: list[TranscriptEvidenceSegmentOut]
    candidates: list[ExtractionCandidateOut]
    candidate_total: int = 0
    candidate_page: int = 1
    candidate_page_size: int = 10
    candidate_total_pages: int = 1
    pending_count: int = 0
    accepted_count: int = 0
    rejected_count: int = 0
    locked_by_current_user: bool = False
    lock_owner_name: str | None = None
    lock_expires_at: datetime | None = None


class CandidateReviewIn(BaseModel):
    review_status: str = Field(pattern="^(accepted|rejected)$")
    original_text: str = Field(min_length=1, max_length=4000)
    clean_text: str = Field(min_length=1, max_length=4000)
    review_note: str = Field(default="", max_length=2000)
    expected_updated_at: datetime | None = None


class CandidateCreateIn(BaseModel):
    source_transcript_segment_id: str | None = None
    original_text: str = Field(min_length=1, max_length=4000)
    clean_text: str = Field(min_length=1, max_length=4000)
    started_at_ms: int = Field(default=0, ge=0)
    ended_at_ms: int = Field(default=0, ge=0)
    review_note: str = Field(default="人工补充遗漏", max_length=2000)

    @model_validator(mode="after")
    def validate_timeline(self):
        if self.ended_at_ms and self.ended_at_ms < self.started_at_ms:
            raise ValueError("结束时间不能早于开始时间")
        return self


class ReviewLeaseOut(BaseModel):
    acquired: bool
    locked_by_current_user: bool
    lock_owner_name: str | None = None
    lock_expires_at: datetime | None = None


class BulkAcceptOut(BaseModel):
    accepted: int
    skipped: int
    skipped_candidate_ids: list[str] = Field(default_factory=list)


class ReviewAudioTicketOut(BaseModel):
    url: str
    expires: int


class ReviewAudioWaveformOut(BaseModel):
    duration_seconds: float
    peaks: list[float] = Field(default_factory=list)


class CandidateRevisionOut(BaseModel):
    id: str
    candidate_id: str
    extraction_job_id: str
    action: str
    actor_id: str | None
    actor_name: str | None = None
    before_snapshot: dict | None
    after_snapshot: dict | None
    created_at: datetime
    model_config = {"from_attributes": True}

File: frontend/src/api/extraction.ts
import apiClient from './client'

export interface ExtractionJob {
  id: string
  session_id: string
  transcript_version_id: string
  status: 'queued' | 'running' | 'retry_wait' | 'reviewing' | 'reviewed' | 'failed' | 'superseded'
  provider: string
  model: string
  extractor_version: string
  prompt_version: string
  generation_no: number
  supersedes_job_id: string | null
  retry_count: number
  max_retries: number
  error_code: string | null
  error_message: string | null
  created_at: string
  started_at: string | null
  completed_at: string | null
  review_lock_user_id: string | null
  review_lock_acquired_at: string | null
  review_lock_expires_at: string | null
}

export interface ExtractionCandidate {
  id: string
  extraction_job_id: string
  source_transcript_segment_id: string | null
  sequence_no: number
  source_type: 'llm' | 'human'
  review_status: 'pending' | 'accepted' | 'rejected'
  raw_asr_text: string
  original_text: string
  clean_text: string
  char_start: number | null
  char_end: number | null
  started_at_ms: number
  ended_at_ms: number
  reviewer_id: string | null
  review_note: string
  reviewed_at: string | null
  updated_at: string
  is_low_risk: boolean
  classifier_version: string | null
  predicted_label: number | null
  predicted_dimension: string | null
  prediction_confidence: number | null
  prediction_probabilities: Record<string, number> | null
  classified_at: string | null
  classification_error: string
  classification_status: 'pending_classification' | 'classified' | 'classified_with_fallback' | 'not_active'
  prediction_source: 'remote_embedding' | 'tfidf_production' | 'tfidf_fallback' | null
}

export interface ExtractionQueueItem {
  session_id: string
  run_id: string | null
  user_id: string
  user_name: string
  username: string
  class_group: string | null
  task_id: string
  task_title: string
  sequence_no: number
  completed_at: string | null
  completed_at_source: 'session_end' | 'run_completed' | 'session_start_fallback'
  transcript_version_no: number | null
  transcript_source: string | null
  asr_status: string
  asr_error_code: string | null
  asr_error_message: string | null
  audio_available: boolean
  job: ExtractionJob | null
  candidate_count: number
  pending_count: number
}

export interface ExtractionQueuePage {
  items: ExtractionQueueItem[]
  total: number
  page: number
  page_size: number
  total_pages: number
  class_groups: string[]
  tasks: Array<{ id: string; title: string }>
  statuses: string[]
}

export interface EvidenceSegment {
  id: string
  segment_no: number | null
  text: string
  started_at_ms: number
  ended_at_ms: number
}

export interface ExtractionReviewDetail {
  session_id: string
  run_id: string | null
  user_id: string
  user_name: string
  username: string
  task_id: string
  task_title: string
  sequence_no: number
  transcript_version_id: string | null
  transcript_version_no: number | null
  transcript_source: string | null
  full_text: string
  audio_available: boolean
  asr_status: string
  asr_error_code: string | null
  asr_error_message: string | null
  job: ExtractionJob | null
  job_history: ExtractionJob[]
  segments: EvidenceSegment[]
  candidates: ExtractionCandidate[]
  candidate_total: number
  candidate_page: number
  candidate_page_size: number
  candidate_total_pages: number
  pending_count: number
  accepted_count: number
  rejected_count: number
  locked_by_current_user: boolean
  lock_owner_name: string | null
  lock_expires_at: string | null
}

export interface ReviewLease {
  acquired: boolean
  locked_by_current_user: boolean
  lock_owner_name: string | null
  lock_expires_at: string | null
}

export interface CandidateRevision {
  id: string
  candidate_id: string
  extraction_job_id: string
  action: string
  actor_id: string | null
  actor_name: string | null
  before_snapshot: Record<string, unknown> | null
  after_snapshot: Record<string, unknown> | null
  created_at: string
}

export interface ReviewAudioTicket {
  url: string
  expires: number
}

export interface ReviewAudioWaveform {
  duration_seconds: number
  peaks: number[]
}

export interface ExtractionBatchRerunResult {
  requested: number
  created: number
  skipped: number
  failed: number
  items: Array<{
    session_id: string
    status: 'created' | 'skipped' | 'failed'
    message: string
    job: ExtractionJob | null
  }>
}

export interface ExtractionJobStatus {
  id: string
  session_id: string
  status: ExtractionJob['status']
  generation_no: number
  retry_count: number
  max_retries: number
  candidate_count: number
  error_code: string | null
  error_message: string | null
  created_at: string
  started_at: string | null
  completed_at: string | null
}

export const extractionApi = {
  listQueue(params: {
    page: number
    page_size: number
    keyword?: string
    class_group?: string
    task_id?: string
    status?: string
  }) {
    return apiClient.get<ExtractionQueuePage>('/research/extraction/queue', { params })
  },
  detail(sessionId: string, jobId?: string, candidatePage = 1, candidatePageSize = 10) {
    return apiClient.get<ExtractionReviewDetail>(`/research/extraction/sessions/${sessionId}`, {
      params: {
        job_id: jobId || undefined,
        candidate_page: candidatePage,
        candidate_page_size: candidatePageSize
      }
    })
  },
  enqueue(sessionId: string) {
    return apiClient.post<ExtractionJob>(`/research/extraction/sessions/${sessionId}/enqueue`)
  },
  rerun(sessionId: string) {
    return apiClient.post<ExtractionJob>(`/research/extraction/sessions/${sessionId}/rerun`)
  },
  batchRerun(sessionIds: string[]) {
    return apiClient.post<ExtractionBatchRerunResult>(
      '/research/extraction/sessions/batch-rerun',
      { session_ids: sessionIds },
      { timeout: 120_000 }
    )
  },
  jobStatus(jobId: string) {
    return apiClient.get<ExtractionJobStatus>(
      `/research/extraction/jobs/${jobId}/status`
    )
  },
  classifyJob(jobId: string) {
    return apiClient.post<ExtractionCandidate[]>(`/research/extraction/jobs/${jobId}/classify`)
  },
  jobStatuses(jobIds: string[]) {
    return apiClient.post<{ items: ExtractionJobStatus[] }>(
      '/research/extraction/jobs/status',
      { job_ids: jobIds }
    )
  },
  review(candidateId: string, data: {
    review_status: 'accepted' | 'rejected'
    original_text: string
    clean_text: string
    review_note: string
    expected_updated_at: string
  }) {
    return apiClient.patch<ExtractionCandidate>(`/research/extraction/candidates/${candidateId}`, data)
  },
  addCandidate(sessionId: string, data: {
    source_transcript_segment_id: string | null
    original_text: string
    clean_text: string
    started_at_ms: number
    ended_at_ms: number
    review_note: string
  }) {
    return apiClient.post<ExtractionCandidate>(`/research/extraction/sessions/${sessionId}/candidates`, data)
  },
  complete(sessionId: string) {
    return apiClient.post<ExtractionJob>(`/research/extraction/sessions/${sessionId}/complete`)
  },
  acquireLock(sessionId: string) {
    return apiClient.post<ReviewLease>(`/research/extraction/sessions/${sessionId}/lock`)
  },
  renewLock(sessionId: string) {
    return apiClient.post<ReviewLease>(`/research/extraction/sessions/${sessionId}/lock/renew`)
  },
  releaseLock(sessionId: string) {
    return apiClient.delete<ReviewLease>(`/research/extraction/sessions/${sessionId}/lock`)
  },
  bulkAcceptLowRisk(sessionId: string) {
    return apiClient.post<{ accepted: number; skipped: number; skipped_candidate_ids: string[] }>(
      `/research/extraction/sessions/${sessionId}/candidates/bulk-accept-low-risk`
    )
  },
  candidateHistory(candidateId: string) {
    return apiClient.get<CandidateRevision[]>(`/research/extraction/candidates/${candidateId}/history`)
  },
  audioTicket(sessionId: string) {
    return apiClient.post<ReviewAudioTicket>(
      `/research/extraction/sessions/${sessionId}/audio-ticket`
    )
  },
  audioWaveform(sessionId: string) {
    return apiClient.get<ReviewAudioWaveform>(
      `/research/extraction/sessions/${sessionId}/audio-waveform`,
      { timeout: 120_000 }
    )
  }
}
